apps/gitlabpipelinestatus/gitlab_pipeline.star

```diff
@@ -0,0 +1,80 @@
+createdPendingSkipped = base64.decode("""iVBORw0KGgoAAAANSUhEUgAAACEAAAAgCAYAAACcuBHKAAABBElEQVRYR+2X4Q2EIAyFZSC2uLiPcQ5z+xi3YCAvktQUU+CVXD1y0b9W/Xh9r6AbOrhcBwxDhAgh7DUY770ZsEMACNAK5IRYtzUrxvga4z1ziNIHSC1ziJonblHipxA8HYcv5mk+U2DdBlp4EtEHgmTXKCHNl9b0NLVjeS87zQ5u6NKsoTruOZUnSsk5Vq+ZutL0LSqB7CutEHzuVCEkFXg7OATqiWv0IwS99GrMXBtMILjsyGrMIZDxzWsQT9SinxxUWl2OPMcVFj2hXX1uTkitlPafr0Dkons7BKqeqRIPhFYBpN7sXwKJbbKLIrTaGuRoSDX/rQRyDKDt3EwJTfu6gPgAs7cWEH3DHkAAAAAASUVORK5CYII=""")
+canceled = base64.decode("""iVBORw0KGgoAAAANSUhEUgAAACEAAAAgCAYAAACcuBHKAAABEUlEQVRYR+2XwRnCIAyFYRmH8OLRNZzDAZzDNTx6cQiXoYoFk34BXvhM5dBem7Y/Ly8P6t0Alx+AwUWIcD2HFow/XcyAPQKQAK1AvhD7Q1mMxz3eM4eofSCpZQ7R8sQqSvwVgk3Hyxd+d8xTYN2GbPgNYpaCh5WiHVK+9E5PF0R43oKbs4MZupY1adXEcypP1CbnvXpN6krpW1UC2Vd6IWjuNCEkFWg7KATqieXofyBSjxfGLLXBBILKjqzGHAKJb1oDeaIx+uyg0uty5DmqsOgJ7epLOSG1Utp/fgJRGt3VIVD1TJXYILQKIPVm/xLI2LJdFKHV1iBHw1yjfTlaP4QSyDEgbedmnkAViyCaYqvaCbHH/oEngJdpAAAAAElFTkSuQmCC""")
+manual = base64.decode("""iVBORw0KGgoAAAANSUhEUgAAACEAAAAgCAYAAACcuBHKAAABBUlEQVRYR+2X4RGDIAyFYY2O1jkco3N0tK6BVkSDF+CF6+txnv416peXlwS9G+DyAzC4FSK8XGjB+CnGMi6PAKQPs0AOiEclx0+8R4eofSCpRYdAan1tiKw7Fl/459EF7DLshr8hNinyYWUohzZfeo3bBRHey4TdZkfWVbVZk7IWnjN5ota+3+wtU1ebvlUlkL3SCyEncBNCU0GWQ0Kgnji3foRINT4Zs1QGCoSUHcmGDoHsEBkDeaLR+tlBpdflyHNSYdUT1uxLc0IrpbZ/fgJRat2/Q6DqUZW4IawKIPG0fwmkbbMtitBaY5Cj4R5jfTkaP4QSyDEgrXOaJ1DFVhBLMCt2BhiwvEkYq+zVAAAAAElFTkSuQmCC""")
+pipeline_dict = dict([("invalid", [failed, "API call failed!"]), ("failed", [failed, "Build failed!"]), ("success", [success, "Success"]), ("running", [running, "Running"]), ("created", [createdPendingSkipped, "Created"]), ("pending", [createdPendingSkipped, "Pending"]), ("skipped", [createdPendingSkipped, "Skipped"]), ("canceled", [canceled, "Canceled"]), ("manual", [manual, "Set to manual"])])
+
+def main(config):
+    token = config.get("api-token")
+    projectId = config.get("project-id")
+    branch = config.get("branch")
+    status = get_pipeline_status(token, projectId, branch)
+    ICON = render.Image(src = pipeline_dict.get(status)[0])
+
+    box = render.Column(
+        children = [
+            render.Box(width = 1, height = 12, color = "#000000"),
+            render.WrappedText(content = pipeline_dict.get(status)[1], font = "tom-thumb"),
+        ],
+    )
+    return render.Root(
+        child = render.Row(
+            children =
+                [ICON, box],
+        ),
+    )
+
+def get_pipeline_status(accesstoken, id, ref):
+    # Set the GitLab API endpoint and access token
+    api_endpoint = "https://gitlab.com/api/v4"
+
+    # Specify the project and branch for which to check the pipeline status
+    pipeline_url = "%s/projects/%s/pipelines?ref=%s&access_token=%s" % (api_endpoint, id, ref, accesstoken)
+    pipeline_data = http.get(pipeline_url)
```

Comment:
Might be a good idea to add some caching to this call. Even if it's just 30s or so it'll help to keep request rates down. No idea what rate limits gitlab have in place though.

When Tidbyt generates preview graphics, it calls the app without configuration. In this case, the app should ideally produce an illustrative default graphic that shows the user what they can expect if they set it up themselves with API tokens all that. If the plan is to use a (secret) default token for this, caching these calls will be very important.

Reply:
I did not cache anything because the pipeline's status can (depending on the circumstances) change very often and even if you did cache it, you would still have to make a call to the api to check for the current status and then the caching would be unnecessary. from what google tells me, the maximum per user is 2000 requests per minute, which would certainly not be filled by a user, so i did not add caching. as it is right now, i do not see a need for caching, because as I said, with every call, the pipeline should be able to update, but if you want me to change it to cache the data, I can do that, just let me know and I can add this feature. a feature for showing the pergentage of the pipeline already built if the status is building is already in planning. I could also add the key to some random repository of mine as the standard so the user doesnt have the standard fail message, please also let me know if you want that.